### ISSUE TITLE
useQuery `onCompleted` in v.3.8.0 doesn't envoke during using `refetch` + `skip: true` option

### ISSUE THREAD

**Issue by av-k** [apollographql/apollo-client#11140]
### Issue Description

When I try to use [useQuery](https://www.apollographql.com/docs/react/data/queries#usequery-api) with the next configuration:
```
  const { loading, error, refetch } = useQuery(GET_LOCATIONS, {
    skip: true,
    fetchPolicy: "network-only",
    notifyOnNetworkStatusChange: true,
    onCompleted
  });
```

I'm waiting for that **onCompleted** callback will be invoked immediately after when the request will be finished successfully. Instead of that **onCompleted** is not invoked as it was in version **3.7.17** 

### Link to Reproduction

https://codesandbox.io/s/apollo-client-3-8-0-oncompleted-issue-h3q79j?file=/src/App.js

### Reproduction Steps

1. View the linked codesandbox reproduction.
2. Compare the code with [the similar codesandbox](https://codesandbox.io/s/apollo-client-3-8-0-oncompleted-issue-alive-7c7c2s?file=/src/App.js) to be sure that the mentioned `options` are really the cause of the issue.

**Comment by alessbell** [comment]
Hi @av-k 👋 

You are correct, this behavior did change in 3.8.0 with https://github.com/apollographql/apollo-client/pull/10229, which fixed a regression that occurred between versions `3.4.17` and `3.5.0` and beyond.

The issue was `onCompleted` firing when cache writes happened without any network request from the associated hook which was admittedly confusing, so in 3.8.0 we reverted this behavior. (It's technically a bug fix, but since it's not backward compatible we avoided shipping it in a patch.)

Can I ask for more information about what you're trying to accomplish by using `skip: true` + `fetchPolicy: "network-only` with `onCompleted` to set data on local component state after calling `refetch` inside a `useEffect` hook? I would find it surprising for a query with `skip: true` to fire `onCompleted` at all since it's by definition not running, but with some more info we may be able to help you implement the behavior you're after.

**Comment by alessbell** [comment]
Also, as a follow up: if you're looking to sync some state like `locations` which is derived from `data.locations`, instead of using `useState` and syncing it in a callback via `onCompleted` (which causes an additional render), consider just using a simple variable instead:

```diff
- const [locations, setLocations] = useState([]);
- const onCompleted = (response) => {
-   setLocations(response.locations || []);
- };
- const { loading, error, refetch } = useQuery(GET_LOCATIONS, {
-   skip: true,
-   fetchPolicy: "network-only", // you may want to also read cache data now
-   notifyOnNetworkStatusChange: true,
-   onCompleted
- });
+ const { data, loading, error } = useQuery(GET_LOCATIONS); 
+ const locations = data?.locations || [];
```

As Dominik ([TkDodo](https://github.com/TkDodo)) of [TanStack/query](https://github.com/TanStack/query) wrote in a great [blog post](https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose#state-syncing), this avoids a few problems: needing to sync state, which may get out of sync if a callback doesn't run as we see with 3.8 fixing our `onCompleted` behavior, and triggering an additional render when setting state.

Hopefully this feels relevant, but let me know if there's something I'm missing here :)

**Comment by vieira** [comment]
> Hi @av-k 👋
> 
> You are correct, this behavior did change in 3.8.0 with #10229, which fixed a regression that occurred between versions `3.4.17` and `3.5.0` and beyond.
> 
> The issue was `onCompleted` firing when cache writes happened without any network request from the associated hook which was admittedly confusing, so in 3.8.0 we reverted this behavior. (It's technically a bug fix, but since it's not backward compatible we avoided shipping it in a patch.)
> 
> Can I ask for more information about what you're trying to accomplish by using `skip: true` + `fetchPolicy: "network-only` with `onCompleted` to set data on local component state after calling `refetch` inside a `useEffect` hook? I would find it surprising for a query with `skip: true` to fire `onCompleted` at all since it's by definition not running, but with some more info we may be able to help you implement the behavior you're after.

This was surprising to me the first time I found about this behaviour but then it turned out to be convenient in some cases, e.g., having a callback being invoked both when an entity is fetched and when a subscription updates (in the cache) that entity.

Is there some clean way to have a callback being invoked whenever the result of a query (that is received as an argument) changes (in the cache)?

**Comment by alessbell** [comment]
Hi @vieira 👋 

> Is there some clean way to have a callback being invoked whenever the result of a query (that is received as an argument) changes (in the cache)?

When any part of the data watched by your `useQuery` call changes, the hook causes your component to re-render with the new `data`, so I'm curious why you'd want to have a separate callback for that. You can always store some derived data in a separate variable as I showed above without having to worry about synchronizing state, and can memoize it with `useMemo` if needed, e.g.:

```js
function SomeComponent({ tab }) {
  const { data } = useQuery(QUERY);
  const visibleTodos = useMemo(
    () => filterTodos(data.todos, tab),
    [data, tab]
  );
  // ...
}
```

Does this meet your requirements? If not, an example would be very helpful in understanding what you're trying to do.

I will add: it's very unlikely we'll be _adding_ a callback here for `onCacheUpdate` or similar, for many of the reasons in [Dominik's post](https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose#state-syncing), among others. I hope the reasoning is clear, even if you don't agree with the outcome :)

**Comment by vieira** [comment]
Hello @alessbell,

It does, thanks! For some reason I thought that a subscription updating the entity in the cache would not make the relevant `useQuery` trigger a rerender with the updated data but it does and it is exactly what I needed.

```diff
const useLiveQuery = function (queryDocument, subscriptionDocument, {
  queryOptions = {},
  subscriptionOptions = {},
  onData,
  ...,
} = {}) {
  const [execute, query] = useLazyQuery(queryDocument, {
-   onCompleted: onData,
    ...queryOptions,
  });

  useSubscription(subscriptionDocument, {
    ...subscriptionOptions,
  });

  // some unrelated logic to execute the query or refetch when websocket reconnects
  // ...

+ const { data } = query;
+
+ useEffect(() => { 
+   if (!data || !onData) return;
+   onData(data);
+ }, [data]);

  return { ...query, loading: !called || loading };
}
```

Thanks! 👍

**Comment by alessbell** [comment]
@vieira great to hear! And happy to help :)

Since this is working as expected, I'll go ahead and close this issue out, but don't hesitate to ping me [on Discord](https://discord.gg/graphos) with any other questions. Thanks!

**Comment by av-k** [comment]
> Hi @av-k 👋
> 
> You are correct, this behavior did change in 3.8.0 with #10229, which fixed a regression that occurred between versions `3.4.17` and `3.5.0` and beyond.
> 
> The issue was `onCompleted` firing when cache writes happened without any network request from the associated hook which was admittedly confusing, so in 3.8.0 we reverted this behavior. (It's technically a bug fix, but since it's not backward compatible we avoided shipping it in a patch.)
> 
> Can I ask for more information about what you're trying to accomplish by using `skip: true` + `fetchPolicy: "network-only` with `onCompleted` to set data on local component state after calling `refetch` inside a `useEffect` hook? I would find it surprising for a query with `skip: true` to fire `onCompleted` at all since it's by definition not running, but with some more info we may be able to help you implement the behavior you're after.

Hi @alessbell!

First of all, thank you for the quick response, and sorry for my long answer.
All of us have our own goals and approaches to how we use @apollo/client.
In my case, I understood the case why it was fixed as it is now.
My point is more about the not obvious documentation.

At least there are two points that have to be noticed in the docs about `useQuery`.
The first, `Result > Operation data > data — An object containing the result of your GraphQL query after it completes.`. Said nothing about that data won't be updated if `useQuery` was initialized with option = `skip: true`.
The second, the fix that has been done in `3.8.0` for `onCompleted` has a similar nature of occurrence, it's the same way how data-flow have been working with `data` along time. But `Options > Operation options > onCompleted — A callback function that's called when your query successfully completes with zero errors (or if errorPolicy is ignore and partial data is returned).` has also nothing mentioned that this callback won't be called if `useQuery` was initialized with option = `skip: true`.

The real case of how I use `useQuery` together with the option `skip: true` is the next.
A CRUD component that takes from URL params an ID of an entity.
If the received param is not ID, so we have a new entity, and the CRUD component will be initialized with predefined initial values, otherwise `useQuery` will be invoked immediately.

To speak more specifically here is an example:
```
import {useParams} from 'react-router-dom';
import {useQuery, gql} from "@apollo/client";

import {SomeInternalStuff} from 'client/components/Stuff';

type Props = {
  sanitizedEntity: unknown,
  handleEntityOnLoad: (entity: unknown) => void,
};

/**
 * Fetch user
 */
const useFetchUser: (props: {variables: {id: string}, skip: boolean, onCompleted: ((data: any) => void) | undefined}) => {user: any, userLoading: boolean} = props => {
  const {variables, skip} = props;
  const {data, loading} = useQuery(
    usersGetUserIdQuery,
    {
      variables,
      skip,
      fetchPolicy: 'network-only',
      notifyOnNetworkStatusChange: true,
      onCompleted,
    },
  );

  return {
    user: data?.user,
    userLoading: loading,
  };
};

const CRUDForm: React.FC<Props> = function CRUDForm({sanitizedEntity, handleEntityOnLoad}) {
  const {id} = useParams();
  const {user, loading} = useFetchUser(GET_USER, {
    variables: {id},
    skip: !id || id === 'add',
    onCompleted: handleEntityOnLoad,
  });

  return <SomeInternalStuff entity={sanitizedEntity} />
};
```

What a pity, this kind of "declarative" logic won't work anymore, It has to be rewritten to only one possible `useLazyQuery` logic way.

@alessbell is it possible to expect that you will make changes to the documentation or change the data-flow logic of `useQuery` when the option `skip` is set to `false` on the initialization step?
